File: ecosystem/simulator.py
from components import Planet, Ecosystem, Prey, Predator
from constants import *

import random
from threading import Lock, Thread


def new_position(ri, ci, grid_shape):
    valid_positions = []
    if ri + 1 < grid_shape[0] and ci + 1 < grid_shape[1]:
        # Bottom right
        valid_positions.append((ri + 1, ci + 1))
    if ri + 1 < grid_shape[0]:
        # Bottom
        valid_positions.append((ri + 1, ci))
    if ri + 1 < grid_shape[0] and ci - 1 >= 0:
        # Bottom left
        valid_positions.append((ri + 1, ci - 1))
    if ci - 1 >= 0:
        # Left
        valid_positions.append((ri, ci - 1))
    if ri - 1 >= 0 and ci - 1 >= 0:
        # Top left
        valid_positions.append((ri - 1, ci - 1))
    if ri - 1 >= 0:
        # Top
        valid_positions.append((ri - 1, ci))
    if ri - 1 >= 0 and ci + 1 < grid_shape[1]:
        # Top right
        valid_positions.append((ri - 1, ci + 1))
    if ci + 1 < grid_shape[1]:
        # Right
        valid_positions.append((ri, ci + 1))

    random_position = random.randint(0, len(valid_positions) - 1)
    return valid_positions[random_position]


def apply_position(target, source):
    target[PREDATORS].clear()
    target[PREY].clear()
    for predator in source[PREDATORS]:
        target[PREDATORS].append(predator)
    for prey in source[PREY]:
        target[PREY].append(prey)


def get_survivors(prey, predators, food):
    surviving_prey = []
    surviving_predators = []
    end_confront_idx = min(len(prey), len(predators))
    for i in range(end_confront_idx):
        current_prey = prey[i]
        current_predator = predators[i]
        if current_prey.confront_day(current_predator.inherited[STRENGTH], food):
            surviving_prey.append(current_prey)
        if current_predator.confront_day(current_prey.inherited[STRENGTH]):
            surviving_predators.append(current_predator)

    for i in range(end_confront_idx, len(prey)):
        current_prey = prey[i]
        if current_prey.confront_day(0, food):
            surviving_prey.append(current_prey)
    for i in range(end_confront_idx, len(predators)):
        current_predator = predators[i]
        if current_predator.confront_day(float("nan")):
            surviving_predators.append(current_predator)
    return surviving_prey, surviving_predators


def reproduce(parents):
    parents_and_offspring = []
    parents_and_offspring.extend(parents)
    if parents:
        correction = 0
        if len(parents) % 2 == 1:
            correction = 1
        for left_idx in range(0, len(parents) - correction, 2):
            left_parent, right_parent = (
                parents[left_idx],
                parents[left_idx + 1],
            )
            offspring = left_parent.reproduce(right_parent)
            parents_and_offspring.extend(offspring)
    return parents_and_offspring


def process_position(ri, ci, day_count):
    for day_num in range(day_count):
        with Planet.worker_state_lock:
            while Planet.worker_state != PROCESS:
                Planet.worker_state_cv.wait()

        with Planet.grid[ri][ci][0][POSITION_LOCK]:
            surviving_prey, surviving_predators = get_survivors(
                Planet.grid[ri][ci][0][PREY],
                Planet.grid[ri][ci][0][PREDATORS],
                Planet.grid[ri][ci][0][FOOD],
            )

            next_prey = reproduce(surviving_prey)
            next_predators = reproduce(surviving_predators)

            for prey in next_prey:
                new_ri, new_ci = new_position(ri, ci, Planet.grid_shape)
                with Planet.grid[new_ri][new_ci][1][POSITION_LOCK]:
                    Planet.grid[new_ri][new_ci][1][PREY].append(prey)
            for predator in next_predators:
                new_ri, new_ci = new_position(ri, ci, Planet.grid_shape)
                with Planet.grid[new_ri][new_ci][1][POSITION_LOCK]:
                    Planet.grid[new_ri][new_ci][1][PREDATORS].append(predator)

        with Planet.ready_lock:
            Planet.ready += 1
            if Planet.ready == Planet.worker_count:
                Planet.ready = 0
                ecosystem_alive = Ecosystem.display_status() > 0
                Prey.display_status()
                Predator.display_status()
                with Planet.worker_state_lock:
                    if ecosystem_alive:
                        Planet.worker_state = TRANSITION
                        Prey.display_evolution_status()
                        Prey.reset_round_status()
                        Predator.display_evolution_status()
                        Predator.reset_round_status()
                    else:
                        Planet.worker_state = DONE
                    Planet.worker_state_cv.notify_all()

        with Planet.worker_state_lock:
            while Planet.worker_state == PROCESS:
                Planet.worker_state_cv.wait()
            if Planet.worker_state == DONE:
                return

        with Planet.grid[ri][ci][0][POSITION_LOCK]:
            with Planet.grid[ri][ci][1][POSITION_LOCK]:
                apply_position(Planet.grid[ri][ci][0], Planet.grid[ri][ci][1])
                apply_position(
                    Planet.grid[ri][ci][1],
                    {
                        PREDATORS: [],
                        PREY: [],
                    },
                )

        with Planet.ready_lock:
            Planet.ready += 1
            if Planet.ready == Planet.worker_count:
                Planet.ready = 0
                with Planet.worker_state_lock:
                    Planet.worker_state = PROCESS
                    Planet.worker_state_cv.notify_all()


def run_simulation():
    # Potential user inputs
    Planet.grid_shape = (10, 10)
    start_predator_count = 20  # can replace with predator types
    start_prey_count = 20  # can replace with prey types
    start_food_source_count = 100  # can replace with food amounts/positions
    day_count = 10
    prey_start_traits = {
        STRENGTH: random.randint(0, 5),
        OFFSPRING_CAPACITY: 4,
        CALORIE_USAGE: 1,
    }
    predator_start_traits = {
        STRENGTH: random.randint(3, 6),
        OFFSPRING_CAPACITY: 4,
        CALORIE_USAGE: 0.25,
    }

    Planet.grid = []
    for ri in range(Planet.grid_shape[0]):
        Planet.grid.append([])
        for ci in range(Planet.grid_shape[1]):
            Planet.grid[-1].append(
                (
                    {
                        PREDATORS: [],
                        PREY: [],
                        FOOD: 0,
                        POSITION_LOCK: Lock(),
                    },
                    {
                        PREDATORS: [],
                        PREY: [],
                        FOOD: 0,
                        POSITION_LOCK: Lock(),
                    },
                )
            )

    for _ in range(start_food_source_count):
        ri, ci = random.randint(0, Planet.grid_shape[0] - 1), random.randint(
            0, Planet.grid_shape[1] - 1
        )
        with Planet.grid[ri][ci][0][POSITION_LOCK]:
            Planet.grid[ri][ci][0][FOOD] += 1
            Planet.grid[ri][ci][1][FOOD] += 1

    for _ in range(start_predator_count):
        ri, ci = random.randint(0, Planet.grid_shape[0] - 1), random.randint(
            0, Planet.grid_shape[1] - 1
        )
        with Planet.grid[ri][ci][0][POSITION_LOCK]:
            predator = Predator(_inherited=predator_start_traits)
            Planet.grid[ri][ci][0][PREDATORS].append(predator)

    for _ in range(start_prey_count):
        ri, ci = random.randint(0, Planet.grid_shape[0] - 1), random.randint(
            0, Planet.grid_shape[1] - 1
        )
        with Planet.grid[ri][ci][0][POSITION_LOCK]:
            prey = Prey(_inherited=prey_start_traits)
            Planet.grid[ri][ci][0][PREY].append(prey)

    Planet.worker_count = Planet.grid_shape[0] * Planet.grid_shape[1]

    assert Ecosystem.display_status() > 0
    Prey.display_status()
    Predator.display_status()
    Prey.display_evolution_status()
    Predator.display_evolution_status()
    # START THREADS
    threads = []
    for ri in range(Planet.grid_shape[0]):
        for ci in range(Planet.grid_shape[1]):
            t = Thread(target=process_position, args=[ri, ci, day_count])
            t.start()
            threads.append(t)
    for thread in threads:
        thread.join()
    Ecosystem.save_status_history()
    Prey.save_status_history()
    Predator.save_status_history()
    Prey.save_evolution_history()
    Predator.save_evolution_history()


if __name__ == "__main__":
    run_simulation()

File: ecosystem/cplusplus/simulator.cpp
#include "organisms.h"
#include "planet.h"
#include "random_numbers.h"
#include "simulation_parameters.h"
#include "simulator.h"
#include "organisms.h"

#include <vector>
#include <iostream>
#include <cstdlib>
#include <thread>
#include <mutex>
#include <algorithm>
#include <utility>
#include <cassert>
#include <stdio.h>

using std::cout;
using std::endl;
using std::vector;


Simulator::Simulator() : ready(0), worker_count(PLANET_GRID_HEIGHT * PLANET_GRID_WIDTH), 
                         worker_state(WorkerState::PROCESS),
                         planet(Planet(PLANET_GRID_HEIGHT, PLANET_GRID_WIDTH)), day_count(DAY_COUNT) {}

void Simulator::wait_for_processing() {
    std::unique_lock<std::mutex> worker_state_lock(worker_state_mutex);
    worker_state_cv.wait(worker_state_lock, [this] { return worker_state == WorkerState::PROCESS; });
}

void Simulator::populate_survivors(const PlanetPositionAccess& pos,
                                vector<Prey*>& surviving_prey,
                                vector<Predator*>& surviving_predators) {
    size_t prey_count = pos.get_prey_ptr()->size();
    size_t predator_count = pos.get_predators_ptr()->size();
    if (predator_count > 0) {
        vector<std::pair<Predator*, vector<Prey*>>> predators_to_prey;
        for (Predator* predator : *pos.get_predators_ptr()) {
            predators_to_prey.push_back(std::make_pair(predator, vector<Prey*>()));
        }
        if (prey_count < predator_count) {
            for (size_t prey_idx = 0; prey_idx < prey_count; ++prey_idx) {
                predators_to_prey[prey_idx].second.push_back((*pos.get_prey_ptr())[prey_idx]);
            }
        }
        else {
            size_t predator_idx = 0;
            for (size_t prey_idx = 0; prey_idx < prey_count; ++prey_idx) {
                if (predator_idx == predator_count) {
                    predator_idx = 0;
                }
                predators_to_prey[predator_idx].second.push_back((*pos.get_prey_ptr())[prey_idx]);
                ++predator_idx;
            }
        }
        for (std::pair<Predator*, vector<Prey*>> predator_prey : predators_to_prey) {
            predator_prey.first->eat_for_day(predator_prey.second);
            if (predator_prey.first->still_alive()) {
                surviving_predators.push_back(predator_prey.first);
            }
            else {
                delete predator_prey.first;
            }
        }
    }

    for (Prey* one_prey : *pos.get_prey_ptr()) {
        if (one_prey->still_alive()) {
            one_prey->eat_for_day(*pos.get_food_ptr());
            if (one_prey->still_alive()) {
                surviving_prey.push_back(one_prey);
            }
            else {
                delete one_prey;
            }
        }
    }
}

template <class T>
void Simulator::reproduce_organisms(std::vector<T*>& organisms) {
    size_t organism_count = organisms.size();
    for (size_t org_right_idx = 1; org_right_idx < organism_count; org_right_idx += 2) {
        organisms[org_right_idx - 1]->reproduce(organisms[org_right_idx], organisms);
    }
}
template void Simulator::reproduce_organisms(std::vector<Prey*>& organisms);
template void Simulator::reproduce_organisms(std::vector<Predator*>& organisms);

std::pair<size_t, size_t> Simulator::get_next_location(size_t row, size_t col) {
    vector<std::pair<size_t, size_t>> valid_positions;
    size_t planet_height = planet.get_height();
    size_t planet_width = planet.get_width();

    valid_positions.push_back({ row, col });
    if (row + 1 < planet_height && col + 1 < planet_width)
        valid_positions.push_back({ row + 1, col + 1 });
    if (row + 1 < planet_height)
        valid_positions.push_back({ row + 1, col });
    if (row + 1 < planet_height && col >= 1)
        valid_positions.push_back({ row + 1, col - 1 });
    if (col >= 1)
        valid_positions.push_back({ row, col - 1 });
    if (row >= 1 && col >= 1)
        valid_positions.push_back({ row - 1, col - 1 });
    if (row >= 1)
        valid_positions.push_back({ row - 1 , col });
    if (row >= 1 && col + 1 < planet_width)
        valid_positions.push_back({ row - 1, col + 1 });
    if (col + 1 < planet_width)
        valid_positions.push_back({ row, col + 1 });

    size_t random_position = random_int(0, valid_positions.size() - 1);
    std::pair<size_t, size_t> new_pos = valid_positions[random_position];
    assert(new_pos.first >= 0 && new_pos.first < planet_height && new_pos.second >= 0 && new_pos.second < planet_width);
    return new_pos;
}

void Simulator::play_out_day(size_t row, size_t col, vector<Prey*>& next_prey, vector<Predator*>& next_predators) {
    PlanetPositionAccess pos = planet.write(row, col);

    populate_survivors(pos, next_prey, next_predators);

    reproduce_organisms(next_prey);
    reproduce_organisms(next_predators);

    pos.get_prey_ptr()->clear();
    pos.get_predators_ptr()->clear();
}

bool Simulator::get_ecosystem_status() {
    return true;
}

void Simulator::set_worker_state(WorkerState next_worker_state, bool check_ecosystem_health) {
    std::unique_lock<std::mutex> ready_count_lock(ready_mutex);
    ++ready;
    if (ready == worker_count) {
        ready = 0;
        if (check_ecosystem_health && !get_ecosystem_status()) {
            next_worker_state = WorkerState::DONE;
        }
        std::unique_lock<std::mutex> worker_state_lock(worker_state_mutex);
        worker_state = next_worker_state;
        worker_state_cv.notify_all();
    }
}

PlanetPositionAccess Simulator::prepare_organism_move(size_t current_row, size_t current_col) {
    std::pair<size_t, size_t> next_row_col = get_next_location(current_row, current_col);
    return planet.write(next_row_col.first, next_row_col.second);
}

bool Simulator::transition_day(size_t row, size_t col, const std::vector<Prey*>& next_prey, const std::vector<Predator*>& next_predators) {
    std::unique_lock<std::mutex> worker_state_lock(worker_state_mutex);
    worker_state_cv.wait(worker_state_lock, [this] { return worker_state != WorkerState::PROCESS; });
    if (worker_state == WorkerState::DONE) {
        return false;
    }
    else if (worker_state == WorkerState::TRANSITION) {
        for (Prey* one_prey : next_prey) {
            PlanetPositionAccess access = prepare_organism_move(row, col);
            access.get_prey_ptr()->push_back(one_prey);
        }
        for (Predator* one_predator : next_predators) {
            PlanetPositionAccess access = prepare_organism_move(row, col);
            access.get_predators_ptr()->push_back(one_predator);
        }
        return true;
    }
    else {
        assert(false);
    }
}

void Simulator::process_position(size_t row, size_t col) {
    for (size_t day = 0; day < day_count; ++day) {
        wait_for_processing();

        vector<Prey*> next_prey;
        vector<Predator*> next_predators;
        play_out_day(row, col, next_prey, next_predators);

        set_worker_state(WorkerState::TRANSITION, true);

        if (transition_day(row, col, next_prey, next_predators)) {
            set_worker_state(WorkerState::PROCESS, false);
        }
        else {
            return;
        }
    }
}

void Simulator::run_simulation(){
    cout << "Initializing simulation\n";
    srand(0);

    for (size_t i = 0; i < START_FOOD_SOURCE_COUNT; ++i) {
        size_t row = random_int(0, PLANET_GRID_HEIGHT - 1);
        size_t col = random_int(0, PLANET_GRID_WIDTH - 1);
        PlanetPositionAccess access = planet.write(row, col);
        *access.get_food_ptr() += 1;
    }

    for (size_t i = 0; i < START_PREY_COUNT; ++i){
        size_t row = random_int(0, PLANET_GRID_HEIGHT - 1);
        size_t col = random_int(0, PLANET_GRID_WIDTH - 1);
        PlanetPositionAccess access = planet.write(row, col);
        access.get_prey_ptr()->push_back(Prey::get_new());
    }

    for (size_t i = 0; i < START_PREDATOR_COUNT; ++i){
        size_t row = random_int(0, PLANET_GRID_HEIGHT - 1);
        size_t col = random_int(0, PLANET_GRID_WIDTH - 1);
        PlanetPositionAccess access = planet.write(row, col);
        access.get_predators_ptr()->push_back(Predator::get_new());
    }
    
    cout << "Starting workers\n";
    vector<std::thread> threads;
    threads.reserve(worker_count);
    for (size_t row = 0; row < PLANET_GRID_HEIGHT; ++row) {
        for (size_t col = 0; col < PLANET_GRID_HEIGHT; ++col) {
            threads.push_back(std::thread(&Simulator::process_position, this, row, col));
        }
    }
    for (std::thread& t : threads) {
        t.join();
    }
    cout << "Simulation complete\n";
}

int main(){
    Simulator simulator;
    simulator.run_simulation();
    return 0;
}
